zmq.go

```diff
@@ -287,43 +294,54 @@ func (c *Conn) readFrame(initialFrame bool) (byte, []byte, error) {
 
 	if flag&2 == 2 {
 		// Long form
-		var buf [8]byte
+		var sizeBuf [8]byte
 		c.conn.SetReadDeadline(time.Now().Add(c.timeout))
-		if _, err := io.ReadFull(c.conn, buf[:8]); err != nil {
+		if _, err := io.ReadFull(c.conn, sizeBuf[:8]); err != nil {
 			return 0, nil, err
 		}
-		for _, b := range buf {
+		for _, b := range sizeBuf {
 			size = (size << 8) | uint64(b)
 		}
 	} else {
 		// Short form
-		var buf [1]byte
+		var sizeBuf [1]byte
 		c.conn.SetReadDeadline(time.Now().Add(c.timeout))
-		if _, err := io.ReadFull(c.conn, buf[:1]); err != nil {
+		if _, err := io.ReadFull(c.conn, sizeBuf[:1]); err != nil {
 			return 0, nil, err
 		}
-		size = uint64(buf[0])
+		size = uint64(sizeBuf[0])
 	}
 
 	if size > MaxBodySize {
 		return 0, nil, errors.New("frame too large")
 	}
 
-	buf := make([]byte, size)
+	// Allocate a buffer large enough to fit the frame if one wasn't
+	// provided.
+	if buf == nil {
+		buf = make([]byte, size)
+	} else if uint64(len(buf)) < size {
+		return 0, nil, fmt.Errorf("buffer of size %v is too small for "+
+			"frame of size %v", len(buf), size)
+	}
+
 	// Prevent timeout during large data read in case of slow connection.
 	c.conn.SetReadDeadline(time.Time{})
-	if _, err := io.ReadFull(c.conn, buf); err != nil {
+	if _, err := io.ReadFull(c.conn, buf[:size]); err != nil {
```

Comment:
the return value below should be truncated to `buf[:size]`, no? otherwise this deviates from the current behavior. it could also be that data beyond `size` isn't zero'd

Reply:
Meant to do that, good catch! Fixed.